libpromises/evalfunction.c

```diff
@@ -1551,6 +1552,44 @@ static FnCallResult FnCallGetMetaTags(EvalContext *ctx, ARG_UNUSED const Policy
 
 /*********************************************************************/
 
+static FnCallResult FnCallBasename(ARG_UNUSED EvalContext *ctx,
+                                   ARG_UNUSED const Policy *policy,
+                                   const FnCall *fp,
+                                   const Rlist *args)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+    if (args == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires a filename as first arg!",
+            fp->name);
+        return FnFailure();
+    }
+
+    char dir[PATH_MAX];
+    strlcpy(dir, RlistScalarValue(args), PATH_MAX);
+    if (dir[0] == '\0')
+    {
+        return FnReturn(dir);
+    }
+
+    char *base = basename(dir);
+
+    if (args->next != NULL)
+    {
+        char *suffix = RlistScalarValue(args->next);
+        int diff = strlen(base) - strlen(suffix);
+        if (diff > 0 && StringSafeEqualN(base + diff, suffix, PATH_MAX))
+        {
+            base[diff] = '\0';
+        }
```

Comment:
`strlen` returns size_t, so there is potential for underflow here. Usually, `int` is 32 bit signed, and `size_t` is 64-bit unsigned. If suffix is intmax bytes longer than base, the code will have serious problems. This is pretty unlikely on normal platforms (4 billion byte files?). Still, better to be safe than sorry. Store the lengths as `size_t` and compare first:

```c
const size_t b_length = strlen(base);
const size_t s_length = strlen(suffix);
if (b_length > s_length && StringSafeEqualN(base + b_length - s_length, suffix, PATH_MAX))
```

Reply:
@karlhto on second thought, this could be made a lot safer and more readable if you use `StringEndsWith()`.

```C
if (StringEndsWith(path, extension))
{
    assert(strlen(path) >= strlen(extension)); // Because StringEndsWith() == true
    path[strlen(path) - strlen(extension)] = '\0'; 
}
```

Yes, there are more passes over the same data than necessary, but safety and readability is more important, IMO.

Reply:
This needs to be slightly different because the suffix should only be removed if it's actually a suffix, but the readability definitely becomes better with the use of `StringEndsWith()`